### ISSUE TITLE
[BUG] EACCES error in spec test

### ISSUE THREAD

**Issue by skpupil** [asterinas/asterinas#670]
435.gromacs, 445.gobmk, 454.calculix, 458.sjeng, 464.h264ref, 471.omnetpp and 482.sphinx3 failed on asterinas:
```
[ERROR]: page fault handler failed: addr: 0x1ffeecd8, err: Error { errno: EACCES, msg: Some("page fault addr is not in current vmar") }
```

473.astar, 481.wrf, 444.namd and 403.gcc  failed on asterinas:
```
[ERROR]: page fault handler failed: addr: 0x103e3948, err: Error { errno: EACCES, msg: Some("page fault addr is not backed up by a vmo") }
```

The file of test item mentioned above: [https://github.com/skpupil/my_spec_test](https://github.com/asterinas/asterinas/issues/url)

**Comment by StevenJiang1110** [comment]
@skpupil, by the way, could you please give some brief introduction about the test suite? Are they a standard test suite to test OS functionalities? If so, how do you like adding these test cases as regression test of Asterinas? So we can run these test cases in GitHub CI.

**Comment by skpupil** [comment]
> @skpupil, could you please give the detailed commands about how to reproduce your results? I look at the [spec_test_item](https://github.com/skpupil/my_spec_test) repo, but don't find the suitable docs about how to run these tests in Asterinas.
435
```
./gromacs_base.amd64-m64-gcc41-nn -silent -deffnm gromacs -nice 0
```
445
```
./gobmk_base.amd64-m64-gcc41-nn --quiet --mode gtp < capture.tst
```
454
```
./calculix_base.amd64-m64-gcc41-nn  -i  beampic
```
458
```
./sjeng_base.amd64-m64-gcc41-nn test.txt
```
464
```
./h264ref_base.amd64-m64-gcc41-nn -d foreman_test_encoder_baseline.cfg
```
471
```
./omnetpp_base.amd64-m64-gcc41-nn omnetpp.ini
```

482
```
./sphinx_livepretend_base.amd64-m64-gcc41-nn ctlfile . args.an4
```
@StevenJiang1110

**Comment by skpupil** [comment]
> @skpupil, by the way, could you please give some brief introduction about the `spec test`? Are they from a standard test suite to test OS functionalities? If so, how do you like adding these test cases as regression test of Asterinas? Then we can run these test cases in GitHub CI.

473
```
./astar_base.amd64-m64-gcc41-nn lake.cfg
```
481
```
./wrf_base.amd64-m64-gcc41-nn
```
444
```
./namd_base.amd64-m64-gcc41-nn --input namd.input --iterations 1 --output namd.out
```
403
```
./gcc_base.amd64-m64-gcc41-nn cccp.i -o cccp.s.1
```

**Comment by skpupil** [comment]
> @skpupil, by the way, could you please give some brief introduction about the `spec test`? Are they from a standard test suite to test OS functionalities? If so, how do you like adding these test cases as regression test of Asterinas? Then we can run these test cases in GitHub CI.

Yes, they are from a standard spec CPU 2006 test suite to test OS functionalities.  Since the test framework,  which is dependent on perl and many OS configs,  could not be run on asterinas, we separated each test item from test suite framework for testing after compilation.

I will add these test cases as regression test of Asterinas after all problems are fixed.

**Comment by cchanging** [comment]
@skpupil , this appears to be a memory-related bug. Currently two obvious bugs are due to the `resize` of the VMO and bugs with copy-on-write during fork. You can try using PR #632 first which fixes the bug of fork.

**Comment by junyang-zh** [comment]
I know why. 0x1ffeecd8 > 500 MiB, which collides with the stack.

Fixed by #679.

**Comment by skpupil** [comment]
Can you tell me the memory layout of aster in detail,  why not just use the same memory layout as linux? Memory system is the basic part of the entire operating system, scheduling module network module and file system are dependent on the memory system to work, I think the memory system structure and api interface stability is fatal important.  According to yesterday's meeting, the progress of the other modules seems to be blocked by the instability of the memory system api. @junyang-zh @StevenJiang1110 @cchanging

**Comment by junyang-zh** [comment]
> Could you tell me the memory layout of aster in detail, why not just use the same memory layout as linux? Memory system is the basic part of the entire operating system, scheduling module network module and file system are dependent on the memory system to work, I think the memory system structure and api interface stability is fatal important. According to yesterday's meeting, the progress of the other modules seems to be blocked by the instability of the memory system api. @junyang-zh @StevenJiang1110 @cchanging

For the original design of memory layout I don't know the rational but I suppose that is a result of a rush achieveing PoC in the early stages of Jinux. I discovered it by GDB. The design of #679 is closer to that of Linux.

We will work harder to develop in a stablized way.

**Comment by StevenJiang1110** [comment]
> Could you tell me the memory layout of aster in detail

@skpupil , the original VM layout can be found in the documentation of `aster-nix/src/process/process_vm/mod.rs`. 

As @junyang-zh mentioned, in the early days of Asterinas, running applications that consume a significant amount of virtual memory was not a priority, so we did not focus on achieving consistency with Linux. However, for applications that require a large amount of virtual memory, the VM layout should be redesigned, as demonstrated in #679.

**Comment by skpupil** [comment]
![image](https://github.com/asterinas/asterinas/assets/56211074/35ba3290-e465-40e0-8ae0-ce547e2668fa)

I test 435 agin. Bug does not disappear.

**Comment by skpupil** [comment]
> ![image](https://private-user-images.githubusercontent.com/56211074/314398279-35ba3290-e465-40e0-8ae0-ce547e2668fa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA5MTcyNzAsIm5iZiI6MTcxMDkxNjk3MCwicGF0aCI6Ii81NjIxMTA3NC8zMTQzOTgyNzktMzViYTMyOTAtZTQ2NS00MGUwLThhZTAtY2U1NDdlMjY2OGZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzIwVDA2NDI1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZhMzU3YzdjZGIzODc1MTE0YTBmMzcxZGYwZTdjMWNkNTM3ZTM3YzJlMTgxZWI2MDAxNTFmMDhiYTM5ZGM0MTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.zxU9RPTmfoAEgEVqhXY2RLWZX7rayUNHY-XreV-Ik08)
> 
> I test 435 agin. Bug does not disappear.

445 is the same. the test program crashed, then the os crashed after about 1s because "rhs <= self.max_multiplier"
@StevenJiang1110 @junyang-zh @cchanging 

![image](https://github.com/asterinas/asterinas/assets/56211074/0190bb0e-14c3-454b-b60f-4a4516ec668e)
